docs/Blog-Post.md

```diff
@@ -0,0 +1,525 @@
+restarts.
+
+# Haliax: Legibility via Named Tensors
+
+Haliax is a library for named tensors, built on JAX and [Equinox](https://github.com/patrick-kidger/equinox),
+which is a neural network library for JAX that provides a familiar, PyTorch-like module structure. Haliax uses
+Equinox's module structure for its neural network library, rather than Flax or Haiku.
+
+Named Tensors are a powerful abstraction that allow you to give names to the axes of your tensors. These names help
+make your code more legible, more composable, and less bug-prone. In Haliax, they also form the basis of how we handle
+scale with [Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/) and tensor parallelism.
+(See below for our tutorial!)
+
+Haliax is modeled on Alexander Rush's [Tensor Considered Harmful](https://nlp.seas.harvard.edu/NamedTensor),
+which argues that named tensors are a better abstraction than the positional axes that are common in deep learning.
+In particular, he argues that:
+
+* Named axes are more semantically meaningful and rely less on bitrot-prone comments.
+* Named axes allow you to abstract over unreferenced dimensions, making code more flexible.
+* Broadcasting leads to unreadable strings of `view`s and `squeeze`s.
+
+To this, I'd add that the implicit broadcasting so common in deep learning code is a source of easy-to-miss bugs, and
+that named tensors eliminate many of these bugs.
+
+## A Quick Example: Attention in Haliax
+
+This blog post isn't the place for a full introduction to Haliax (please see the [Haliax tutorial](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC)),
+but here's a quick example of a minimal, but full-featured, attention implementation in Haliax.
+
+```python
+import jax.numpy as jnp
+import haliax as hax
+
+# Named Axes for Tensor Dimensions
+Pos = hax.Axis("position", 1024)  # sequence
+KPos = Pos.alias("key_position")  # key sequence for attention
+Head = hax.Axis("head", 8)  # number of attention heads
+Key = hax.Axis("key", 64)  # key/query/value size
+Embed = hax.Axis("embed", 512)  # embedding size
+
+def attention(Key, KPos, query, key, value, mask):
+    # how similar is each query to each key
+    scores = hax.dot(Key, query, key) / jnp.sqrt(Key.size)
+
+    # mask out invalid positions
+    if mask is not None:
+      scores -= 1E9 * (1.0 - mask)
+
+    # convert to probabilities
+    scores = hax.nn.softmax(scores, axis=KPos)
+
+    # weighted sum of values
+    return hax.dot(KPos, scores, value)
+```
+
+In this example, we've defined an axis for each dimension of our tensors. In Haliax, the named `Axis` is the basic
+building block of named tensors, pairing a name with a size. We can then use these axes to define our tensors, and use
+those axes to perform operations like `softmax` and tensor multiplication (`dot`).
+
+## Compositionality
```

Comment:
Not sure I'd call this composibility / modularity...  seems like more like automatic generalization or something...

Reply:
yeah it's more abstraction...